_posts/2018-04-06-middleware.md

```diff
@@ -0,0 +1,207 @@
+---
+title: Manipulate events using a middleware
+author: Georgos Landin Xanthopoulos
+date: '2018-02-02 15:06:47 +0200'
+header:
+  teaser: images/posts/routelarge.jpg
+  caption: "Photo credit: [**Flickr**](https://www.flickr.com/photos/rockys_photos/355216949/in/photolist-6enFxo-KGos4s-bziePy-91523N-TpGWrG-dvMSg3-6ZREpd-xozCa-3CVfrH-4r4Dz7-pkqvKY-5CL4FN-fHW3WW-8vvEYu-4KfNRw-6xwQuD-6xHB5q-8xo6sb-7dphxn-7KgWrr-342fEp-6xX2tf-6dBTjp-9nB8TA-8xk5pz-P61pjV-dNXAvC-5n8o7A-u7GL9y-6jzRhi-MGt7o1-7A7jLS-i4tC8C-4hDeBQ-3eX22G-dvGvj4-7A3gUD-5b6sLw-5yjxW2-8hxfmE-4cTxKp-Cq8sKM-tQg5e-7A7m1d-6ygJN3-avLA1p-9mQJiX-8ZtQtV-rpFoVZ-7A1DLZ)"
+categories:
+- Symfony
+---
+
+Events are great due to the fact that they are very versatile and can be very useful in most cases.
+Maybe you want to log something specific in your system, or maybe you are building a gamification application that gives
+users experience points based on their actions. Then events might just be right for you.
+
+This post will cover how you can add additional data to your events using the simple bus middleware.
+## How it works
+Classes needed:
+* Middleware - dispatches the ExtractSimpleBusMessage event
+* ExtractSimpleBusMessage - uses an event and a DataContainer class
+* DataContainer - modifies the array given to it, through multiple available functions.
+* A decorator class for each event - adds data to the event it's assigned to
+
+In most cases, your middleware will accumulate a number of events, also known as messages. When your application terminates
+, the middleware will dispatch a SimpleBusMessage event for each accumulated message.
+The decorator classes are event subscribers and listen to the ExtractSimpleBusMessage event, they check the simple bus message
+for their assigned event, and if there is a match, they add some more data to it.
+This was just a quick summary of the process, further down you will see the implementation.
+
+## The middleware
+Down below you can see the middleware class. As you can see, it is an event subscriber that dispatches an ExtractSimpleBusMessage
```

Comment:
"event subscriber that dispatches an ExtractSimpleBusMessage" 

This is correct but it sounds way wrong. Maybe you should hide the fact that it is an eventsubscriber for now. You can talk about that at the end. ".... We can optimize this by listening to kernel.terminate...."